test: migrate from tap to node:test and c8

R-Campbell opened the pull request:
A few decisions while converting tests:
* `t.same` and `t.match` were converted to `t.assert.deepEqual`.
* `t.strictSame` was converted to `t.assert.deepStrictEqual`.
* `t.notOk` was converted to `t.assert.equal` comparing the coerced value (using `!!`) to `false`. We could have done `t.assert.ok(${value} === false)` but using `equal` with `false` felt more in the spirit of checking that for things that are not ok.
* `t.pass` was converted to `t.assert.ok(true, ${originalMessage})`
* Snapshots were removed and copied into the appropriate tests.

#### Checklist

- [x] run `npm run test`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

R-Campbell committed:
fix: migrate from tap to node:test and c8

dancastillo commented on .gitignore:
```diff
@@ -148,8 +148,5 @@ yarn.lock
 .vscode
 .idea
 
-#tap files
-.tap/
-
```
Might be worth keeping this around for a bit just in case for people who ran tests

dancastillo reviewed (commented):
Can you update `t.assert.equal` to `t.assert.strictEqual` and `t.assert.deepEqual` to `t.assert.deepStrictEqual`

R-Campbell committed:
update equal to strictEqual and deepEqual to deepStrictEqual

R-Campbell committed:
change ok(false, ...) to fail

R-Campbell commented:
I made that update but ran into an issue with the Symbol that `json-schema-resolver` adds to the schema (specifically [json-schema-resolver.ignore](https://github.com/Eomm/json-schema-resolver/blob/main/ref-resolver.js)) and couldn't find a great way to reference that in the testing code. Instead I used `JSON.parse(JSON.stringify(...))` to strip the Symbols from the schema objects where needed (see [here](https://github.com/fastify/fastify-swagger/pull/843/files#diff-fc2b8af6daf81b33b3249e57b8ce40fcb66bdf3934c48e8aa67951c5dc47fed1R376) for an example). I don't know if that's the best solution, but it works.

I also found some places where `null` needed to be updated to `undefined` (see [here](https://github.com/fastify/fastify-swagger/pull/843/files#diff-33e69073fcabe579eb6d77ec4f8949b3b03931c5158f8b84c1269a6fe933bc28L452) for an example) as well as `t.fail` calls that I missed and changed to `t.assert.fail`.

dancastillo commented on test/mode/static.js:
```diff
@@ -185,7 +203,7 @@ test('non-object specification.document throws an error', async (t) => {
   const fastify = new Fastify()
 
-  t.rejects(async () => await fastify.register(fastifySwagger, config), new Error('specification.document is not an object'))
+  t.assert.rejects(async () => await fastify.register(fastifySwagger, config), new Error('specification.document is not an object'))
```
```suggestion
  await t.assert.rejects(async () => await fastify.register(fastifySwagger, config), new Error('specification.document is not an object'))
```

dancastillo commented on test/spec/openapi/route.js:
```diff
@@ -158,8 +158,8 @@ test('route options - produces', async (t) => {
-  t.ok(definedPath)
-  t.match(definedPath.responses[200].content, {
+  t.assert.ok(definedPath)
+  t.assert.deepEqual(JSON.parse(JSON.stringify(definedPath.responses[200].content)), {
```
this works without `JSON.parse(JSON.stringify(...)`

dancastillo commented on test/spec/openapi/route.js:
```diff
@@ -233,8 +233,8 @@ test('parses form parameters when all api consumes application/x-www-form-urlenc
-  t.ok(definedPath)
-  t.match(definedPath.requestBody.content, {
+  t.assert.ok(definedPath)
+  t.assert.deepEqual(JSON.parse(JSON.stringify(definedPath.requestBody.content)), {
```
this works without `JSON.parse(JSON.stringify(...)`

dancastillo reviewed (commented):
Lets removed the `!!` to do a strict assert of `undefined` like this  `t.assert.strictEqual(!!value, undefined)`t.assert.strictEqual(value, undefined)` this will help explain behavior IMHO

dancastillo commented:
> I made that update but ran into an issue with the Symbol that `json-schema-resolver` adds to the schema (specifically [json-schema-resolver.ignore](https://github.com/Eomm/json-schema-resolver/blob/main/ref-resolver.js)) and couldn't find a great way to reference that in the testing code. Instead I used `JSON.parse(JSON.stringify(...))` to strip the Symbols from the schema objects where needed (see [here](https://github.com/fastify/fastify-swagger/pull/843/files#diff-fc2b8af6daf81b33b3249e57b8ce40fcb66bdf3934c48e8aa67951c5dc47fed1R376) for an example). I don't know if that's the best solution, but it works.
> 
> I also found some places where `null` needed to be updated to `undefined` (see [here](https://github.com/fastify/fastify-swagger/pull/843/files#diff-33e69073fcabe579eb6d77ec4f8949b3b03931c5158f8b84c1269a6fe933bc28L452) for an example) as well as `t.fail` calls that I missed and changed to `t.assert.fail`.

@Eomm wdyt here - `JSON.parse(JSON.stringify(...))` does remove the symbol from the object. Wondering if there is a preferred way to handle this?

R-Campbell committed:
Update test/mode/static.js

Co-authored-by: Dan Castillo <dan.castillo@jasper.ai>
Signed-off-by: Robert Campbell <R-Campbell@users.noreply.github.com>

R-Campbell committed:
remove unneeded parse/stringify

R-Campbell committed:
Merge branch 'master' into convert-tap-to-node-test

Signed-off-by: Robert Campbell <R-Campbell@users.noreply.github.com>

R-Campbell committed:
update strictEqual comparisons

change from coerced value comparison against false to actual value comparison against undefined where necessary

R-Campbell committed:
Merge branch 'convert-tap-to-node-test' of github.com:R-Campbell/fastify-swagger into convert-tap-to-node-test

dancastillo reviewed (approved):
LGTM

gurgunday reviewed (approved):
lgtm
